### PR TITLE
Fix Tor dns leak

### PR DESCRIPTION
**Pull request by darkdh**
Resolves https://github.com/brave/brave-browser/issues/13527

This PR

Disable CNAME adblock for Tor to prevent it from sending DoH or standard DNS query which will not be routed through Tor

~1. Fixes DoH requests should respect proxy config, matching the behavior as Firefox, so that DoH requests will route through Tor
2. Enforce DoH for DNS queries sent by CNAME adblock because users can turn DoH off and we currently don't have a way to route insecure DNS lookup through Tor(proxy config). We can consider implementing `Proxy DNS when using SOCKS v5`(network.proxy.socks_remote_dns) that Firefox has so we don't have to enforce DoH for this
3. Skip CNAME adblock DNS lookup if DoH mode is not secure or DoH server list is empty~

## Submitter Checklist:

- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed).
- [ ] Requested a security/privacy review as needed.

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### DoH off 
1. Follow the section `Using the (Pre)-Master-Secret` to setup `SSLKEYLOGFILE`
https://wiki.wireshark.org/TLS
2. After Wireshark is capturing traffic, use the `dns` filter
3. Launch Brave
4. Go to brave://settings/security and turn off `Use secure DNS`
5. Open Tor window
6. Navigate to https://tools.ietf.org/
7. After site loaded, stop Wireshark capturing
8. There shouldn't be any DNS query or DoH query for `tools.ietf.org`

### DoH enabled
1. Follow the section `Using the (Pre)-Master-Secret` to setup `SSLKEYLOGFILE`
https://wiki.wireshark.org/TLS
2. After Wireshark is capturing traffic, use the `dns` filter
3. Launch Brave
4. Go to brave://settings/security and make sure `Use secure DNS` is on and use custom Cloudflare resolver
5. Open Tor window
6. Navigate to https://tools.ietf.org/
7. After site loaded, stop Wireshark capturing
8. There shouldn't be any DNS query or DoH query for `tools.ietf.org`

### CODE REVIEW

File: browser/net/brave_ad_block_tp_network_delegate_helper.cc [hunk 1]
```diff
@@ -148,13 +161,35 @@ class AdblockCnameResolveHostClient : public network::mojom::ResolveHostClient {
     // See https://crbug.com/872665
     optional_parameters->source = net::HostResolverSource::DNS;
```
Comment (darkdh) [inline browser/net/brave_ad_block_tp_network_delegate_helper.cc:162]: `ResolveHostParameters::secure_dns_mode_override` for `SECURE` mode is useless without specifying doh_server.
So we have to do it through `DnsConfigOverrides`

File: browser/net/brave_ad_block_tp_network_delegate_helper.cc [hunk 2]
```diff
@@ -148,13 +161,35 @@ class AdblockCnameResolveHostClient : public network::mojom::ResolveHostClient {
...
 
+    net::DnsConfigOverrides dns_config_overrides;
+    if (context->IsTor()) {
+      // Enforce DoH for Tor
+      // TODO(darkdh): we can consider implementing
+      // "Proxy DNS when using SOCKS v5" (network.proxy.socks_remote_dns)
+      // like Firefox has so that we don't have to enforce DoH
```
Comment (antonok-edm) [inline browser/net/brave_ad_block_tp_network_delegate_helper.cc:169]: Just a heads up about https://github.com/brave/brave-browser/issues/13414 - we should be able to remove most of this machinery once we rebase onto CR89, so I'd recommend keeping the fixes here as minimal as possible!

**Review by antonok-edm — APPROVED** [review]

**Review by diracdeltas — CHANGES_REQUESTED** [review]
this is great but i don't think we should enable DoH for users who haven't picked a DoH provider, because IIUC they need to consent to the provider's terms of service. for users who haven't picked a DoH provider, can we just disable cname adblocking for them in Tor mode?

File: patches/net-dns-dns_transaction.cc.patch [hunk 3]
```diff
@@ -0,0 +1,14 @@
+diff --git a/net/dns/dns_transaction.cc b/net/dns/dns_transaction.cc
+index c12f1c728de558b236a0144179bfa268da90fb04..685ec6c86d16f480dd4a995b0fb64c01853a5113 100644
+--- a/net/dns/dns_transaction.cc
++++ b/net/dns/dns_transaction.cc
+@@ -405,8 +405,7 @@ class DnsHTTPAttempt : public DnsAttempt, public URLRequest::Delegate {
+     request_->SetExtraRequestHeaders(extra_request_headers);
+     // Disable secure DNS for any DoH server hostname lookups to avoid deadlock.
+     request_->SetDisableSecureDns(true);
+-    request_->SetLoadFlags(request_->load_flags() | LOAD_DISABLE_CACHE |
```
Comment (bridiver) [inline patches/net-dns-dns_transaction.cc.patch:9]: I think you can avoid this patch by defining `LOAD_BYPASS_PROXY` as `LOAD_DISABLE_CACHE`. If that doesn't work for some reason you should use a define/chromium_src override after the original call instead of changing it to minimize the patch

File: browser/net/brave_ad_block_tp_network_delegate_helper.cc (hunk 2 above)
Comment (bridiver) [reply browser/net/brave_ad_block_tp_network_delegate_helper.cc:169]: It seems like we should be setting this stuff when the Tor connection is created because presumably we need to always use it? Not just for this?

Comment (darkdh) [reply browser/net/brave_ad_block_tp_network_delegate_helper.cc:169]: Normally, when we are in Tor window, we don't need to send ANY dns queries. We route all the HTTP/HTTPS traffic to tor proxy without knowing the IP, that means we just send hostname to proxy and the Tor exit node will do the resolving.

File: patches/net-dns-dns_transaction.cc.patch (hunk 3 above)
Comment (darkdh) [reply patches/net-dns-dns_transaction.cc.patch]: fixed

File: browser/net/brave_ad_block_tp_network_delegate_helper_unittest.cc [hunk 4]
```diff
@@ -25,10 +71,40 @@ TEST(BraveAdBlockTPNetworkDelegateHelperTest, NoChangeURL) {
...
-TEST(BraveAdBlockTPNetworkDelegateHelperTest, EmptyRequestURL) {
+TEST_F(BraveAdBlockTPNetworkDelegateHelperTest, EmptyRequestURL) {
   auto request_info = std::make_shared<brave::BraveRequestInfo>(GURL());
   int rc =
       OnBeforeURLRequest_AdBlockTPPreWork(ResponseCallback(), request_info);
   EXPECT_TRUE(request_info->new_url_spec.empty());
   EXPECT_EQ(rc, net::OK);
 }
+
+TEST_F(BraveAdBlockTPNetworkDelegateHelperTest, ByassTor) {
```
Comment (iefremov) [inline browser/net/brave_ad_block_tp_network_delegate_helper_unittest.cc:82]: ByPassTor

File: browser/net/brave_ad_block_tp_network_delegate_helper_unittest.cc [hunk 5]
```diff
@@ -11,12 +11,58 @@
...
+class TestTorBrowserContext : public content::TestBrowserContext {
+ public:
+  TestTorBrowserContext() : content::TestBrowserContext() {}
+  ~TestTorBrowserContext() override = default;
+
+  bool IsTor() const override { return true; }
+};
+}  // namespace
+
+class BraveAdBlockTPNetworkDelegateHelperTest : public testing::Test {
```
Comment (iefremov) [inline browser/net/brave_ad_block_tp_network_delegate_helper_unittest.cc:37]: this fixture is not related to existing tests, so you could leave them as is and give the test class more relevant name

File: chromium_src/net/dns/dns_transaction.cc [hunk 6]
```diff
@@ -0,0 +1,10 @@
+/* Copyright (c) 2021 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "net/base/load_flags.h"
+
+#define LOAD_BYPASS_PROXY LOAD_DISABLE_CACHE
```
Comment (iefremov) [inline chromium_src/net/dns/dns_transaction.cc:8]: I really don't think we should do this for all requests. I checked git history around this code (https://chromium-review.googlesource.com/c/chromium/src/+/1843334/10/net/dns/dns_transaction.cc), and the authors say LOAD_BYPASS_PROXY is here to prevent deadlocks.

So we should either try upstreaming this or limit to tor-related requests

Comment (darkdh) [reply chromium_src/net/dns/dns_transaction.cc:8]: That is for `SetDisableSecureDns`
It's been there since the first DoH commit 
https://chromium-review.googlesource.com/c/chromium/src/+/710554/40/net/dns/dns_transaction.cc#431

File: browser/net/brave_ad_block_tp_network_delegate_helper_unittest.cc (hunk 4 above)
Comment (bridiver) [reply browser/net/brave_ad_block_tp_network_delegate_helper_unittest.cc:82]: actually I think this is a bad name because it makes it sound like it should be bypassing Tor and that's exactly the opposite of what we are doing. Maybe something like DisableInsecureRequestsOverTor

File: chromium_src/net/dns/dns_transaction.cc (hunk 6 above)
Comment (bridiver) [reply chromium_src/net/dns/dns_transaction.cc:8]: yea, I actually can't really imagine why we'd ever want to bypass the proxy for DoH requests. That seems like it would actually prevent requests from working at all if the proxy was required for access and non-proxy requests are blocked by firewall

File: browser/net/brave_ad_block_tp_network_delegate_helper_unittest.cc (hunk 4 above)
Comment (darkdh) [reply browser/net/brave_ad_block_tp_network_delegate_helper_unittest.cc]: fixed

File: browser/net/brave_ad_block_tp_network_delegate_helper_unittest.cc (hunk 5 above)
Comment (darkdh) [reply browser/net/brave_ad_block_tp_network_delegate_helper_unittest.cc]: fixed

File: chromium_src/net/dns/dns_transaction.cc (hunk 6 above)
Comment (darkdh) [reply chromium_src/net/dns/dns_transaction.cc:8]: There is a discussion with upstream net dev
https://groups.google.com/u/1/a/chromium.org/g/net-dev/c/GI4E9ERwPgw

**Comment by darkdh** [comment]
per discussion on slack with @bridiver and @iefremov, we came to a conclusion that disabling CNAME adblock for Tor would be best option now. Considering in order to make DoH route through Tor, we need to remove `LOAD_BYPASS_PROXY` for dns transaction but it might introduce dns and proxy code looping when we need to resolve proxy name.  see (https://groups.google.com/u/1/a/chromium.org/g/net-dev/c/GI4E9ERwPgw)
So if we want to route DoH through Tor, we need fine grained solution for it and also need a bigger discussion about this with privacy and security

**Comment by fmarier** [comment]
> we came to a conclusion that disabling CNAME adblock for Tor would be best option now. Considering in order to make DoH route through Tor, we need to remove LOAD_BYPASS_PROXY for dns transaction but it might introduce dns and proxy code looping when we need to resolve proxy name.

Sounds like a good compromise to me.

File: chromium_src/net/dns/dns_transaction.cc (hunk 6 above)
Comment (darkdh) [reply chromium_src/net/dns/dns_transaction.cc]: proxy hostname resolution will go through DoH if DoH is enabled
Screen Shot 2021-02-03 at 16 19 24

File: browser/net/brave_ad_block_tp_network_delegate_helper.cc [hunk 7]
```diff
@@ -223,6 +221,12 @@ int OnBeforeURLRequest_AdBlockTPPreWork(const ResponseCallback& next_callback,
     return net::OK;
   }
 
+  DCHECK(ctx->browser_context);
+  // DoH or standard DNS quries won't be routed through Tor, so we need to skip
+  // it.
+  if (ctx->browser_context->IsTor()) {
+    return net::OK;
```
Comment (iefremov) [inline browser/net/brave_ad_block_tp_network_delegate_helper.cc:228]: This is going to disable adblocker in tor windows. What we need to do here is an adblocking pass without CNAME resolving, e.g. we can directly call `ShouldBlockAdWithOptionalCname` with empty CNAME

Comment (darkdh) [reply browser/net/brave_ad_block_tp_network_delegate_helper.cc]: fixed

**Review by iefremov — APPROVED** [review]

**Review by diracdeltas — APPROVED** [review]
